api/sandbox.go

```diff
@@ -0,0 +1,114 @@
+package api
+
+import (
+	"context"
+	"syscall"
+
+	"github.com/hyperhq/runv/api"
+	ocispecs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type ContainerConfig struct {
+	Id string
+
+	UGI    *UserGroupInfo
+	Mounts map[string]Mount
+
+	OciSpec ocispecs.Spec
+}
+
+type Mount struct {
+	StorageID string
+	Path      string
+	ReadOnly  bool
+}
+
+type UserGroupInfo struct {
+	User             string
+	Group            string
+	AdditionalGroups []string
+}
+
+type ProcessConfig struct {
+	Container string
+	Id        string
+	UGI       *UserGroupInfo
+
+	OciProcess ocispecs.Process
+}
+
+type Sandbox struct {
+	VM    VM
+	Agent Agent
+
+	Devices map[string]AddedDevice
+
+	Containers map[string]Container
+}
+
+type SandboxBuider struct {
+	VMBuilder
+	AgentBuider
+	Persistence
+	netowrks map[string]Network
+}
+
+func CreateSandbox(SandboxBuider, VMConfig, AgentConfig, SandboxNetworkConfig) *Sandbox {
+
+}
+
+// most sandbox api will call into vm api, and them agent api
+func (sandbox *Sandbox) Start(b *BootConfig) (err error) {
+}
+func (sandbox *Sandbox) Release() ([]byte, error) {
+}
+func (sandbox *Sandbox) SanboxDone() context.Context {
+}
+func (sandbox *Sandbox) Shutdown() api.Result {
+}
+func (sandbox *Sandbox) Destory() {
+}
+func (sandbox *Sandbox) UpdateRoute( /*todo*/ ) error {
+}
+func (sandbox *Sandbox) AddNic( /*todo*/ ) error {
+}
+func (sandbox *Sandbox) AllNics() []*InterfaceConfig {
+}
+func (sandbox *Sandbox) DeleteNic(id string) error {
+}
+func (sandbox *Sandbox) UpdateNic( /*todo*/ ) error {
+}
+func (sandbox *Sandbox) SetCpus(cpus int) error {
+}
+func (sandbox *Sandbox) SetMem(totalMem int) error {
+}
+
+func (sandbox *Sandbox) AddStorage(s Storage) error {
+}
+func (sandbox *Sandbox) RemoveStorage(name string) error {
+}
+func (sandbox *Sandbox) AddContainer(c *ContainerConfig) error {
+}
+func (sandbox *Sandbox) RemoveContainer(id string) error {
+}
+func (sandbox *Sandbox) StartContainer(id string) error {
+}
+func (sandbox *Sandbox) AddProcess(process *ProcessConfig) error {
+}
+func (sandbox *Sandbox) WaitProcess(container, process string) int {
+}
+func (sandbox *Sandbox) SignalProcess(container, process string, signal syscall.Signal) error {
+}
+func (sandbox *Sandbox) TerminalResize(containerId, execId string, row, column int) error {
+}
+
+func (sandbox *Sandbox) Stats() *SandboxStats {
+}
+func (sandbox *Sandbox) ContainerList() []string {
+}
+func (sandbox *Sandbox) Pause() error {
+}
+func (sandbox *Sandbox) Resume() error {
+}
+func (sandbox *Sandbox) Save(path string) error {
+}
```

Comment:
The virtcontainers containers and pod API implement a large part of the above API:

#### `CreatePod`
```Go
// CreatePod is the virtcontainers pod creation entry point.
// CreatePod creates a pod and its containers. It does not start them.
func CreatePod(podConfig PodConfig) (VCPod, error)
```

#### `DeletePod`
```Go
// DeletePod is the virtcontainers pod deletion entry point.
// DeletePod will stop an already running container and then delete it.
func DeletePod(podID string) (VCPod, error)
```

#### `StartPod`
```Go
// StartPod is the virtcontainers pod starting entry point.
// StartPod will talk to the given hypervisor to start an existing
// pod and all its containers.
func StartPod(podID string) (VCPod, error)
```

#### `StopPod`
```Go
// StopPod is the virtcontainers pod stopping entry point.
// StopPod will talk to the given agent to stop an existing pod
// and destroy all containers within that pod.
func StopPod(podID string) (VCPod, error)
```

#### `RunPod`
```Go
// RunPod is the virtcontainers pod running entry point.
// RunPod creates a pod and its containers and then it starts them.
func RunPod(podConfig PodConfig) (VCPod, error)
```

#### `ListPod`
```Go
// ListPod is the virtcontainers pod listing entry point.
func ListPod() ([]PodStatus, error)
```

#### `StatusPod`
```Go
// StatusPod is the virtcontainers pod status entry point.
func StatusPod(podID string) (PodStatus, error)
```

#### `PausePod`
```Go
// PausePod is the virtcontainers pausing entry point which pauses an
// already running pod.
func PausePod(podID string) (VCPod, error)
```

#### `ResumePod`
```Go
// ResumePod is the virtcontainers resuming entry point which resumes
// (or unpauses) and already paused pod.
func ResumePod(podID string) (VCPod, error)
```

#### `CreateContainer`
```Go
// CreateContainer is the virtcontainers container creation entry point.
// CreateContainer creates a container on a given pod.
func CreateContainer(podID string, containerConfig ContainerConfig) (VCPod, VCContainer, error)
```

#### `DeleteContainer`
```Go
// DeleteContainer is the virtcontainers container deletion entry point.
// DeleteContainer deletes a Container from a Pod. If the container is running,
// it needs to be stopped first.
func DeleteContainer(podID, containerID string) (VCContainer, error)
```

#### `StartContainer`
```Go
// StartContainer is the virtcontainers container starting entry point.
// StartContainer starts an already created container.
func StartContainer(podID, containerID string) (VCContainer, error)
```

#### `StopContainer`
```Go
// StopContainer is the virtcontainers container stopping entry point.
// StopContainer stops an already running container.
func StopContainer(podID, containerID string) (VCContainer, error)
```

#### `EnterContainer`
```Go
// EnterContainer is the virtcontainers container command execution entry point.
// EnterContainer enters an already running container and runs a given command.
func EnterContainer(podID, containerID string, cmd Cmd) (VCPod, VCContainer, *Process, error)
```

#### `StatusContainer`
```Go
// StatusContainer is the virtcontainers container status entry point.
// StatusContainer returns a detailed container status.
func StatusContainer(podID, containerID string) (ContainerStatus, error)
```

#### `KillContainer`
```Go
// KillContainer is the virtcontainers entry point to send a signal
// to a container running inside a pod. If all is true, all processes in
// the container will be sent the signal.
func KillContainer(podID, containerID string, signal syscall.Signal, all bool) error
```

#### `ProcessListContainer`
```Go
// ProcessListContainer is the virtcontainers entry point to list
// processes running inside a container
func ProcessListContainer(podID, containerID string, options ProcessListOptions) (ProcessList, error)
```

Reply:
@sameo  These APIs are just wrappers over Sandbox & Persistence APIs, there are too much oci-runtime specific. What we need is just to expose the Sandbox & Persistence APIs. (The Sandbox APIs are the APIs that take Sandbox as the receiver, not  sandbox name).